hw1

Mur0riki opened the pull request.

Mur0riki committed:
hw1

Mur0riki committed:
hw1 with buld/checkstyle/integration tests

Mur0riki committed:
edit pom, fixed test dependency

Mur0riki committed:
edit bot.yml

Mur0riki committed:
add MessageServiceTest
delete CommandConfig.java
delete CommandHandler

CbaKKa commented on bot/src/main/java/edu/java/bot/commands/CommandHelp.java:
```diff
@@ -0,0 +1,46 @@
+        var commandListString = new StringBuilder();
+
+        commandListString.append(STRING_COMMANDS_BOT);
+        listCommand.forEach(command ->
```
Я сделал бы это немного по-другому:

```java
        listCommand.stream()
                   .map(command -> STRING_COMMANDS_ENUMERATION.formatted(command.command(), command.description()))
                   .forEach(commandListString::append)
```

CbaKKa commented on bot/src/main/java/edu/java/bot/commands/CommandList.java:
```diff
@@ -0,0 +1,57 @@
+
+        var userOptional = userService.findUserById(chatId);
+
+        if (userOptional.isPresent()) {
```
Не совсем корректно так использовать `Optional`, ведь одно из его преимуществ в том, что можно не делать проверку на `null`. Попробуй поиграться с `.map()` и `.flatMap()`

CbaKKa commented on bot/src/main/java/edu/java/bot/commands/CommandList.java:
```diff
@@ -0,0 +1,57 @@
+    private String prepareListSitesMessage(List<URI> uriList) {
+        var sitesString = new StringBuilder();
+
+        sitesString.append(USER_TRACK_SITES_MESSAGE.formatted(uriList.size()));
```
Попробуй сделать через Stream API

CbaKKa commented on bot/src/main/java/edu/java/bot/commands/CommandStart.java:
```diff
@@ -0,0 +1,50 @@
+    private String registerUser(long chatId) {
+        Optional userOptional = userService.findUserById(chatId);
+
+        if (userOptional.isEmpty()) {
```
Тоже самое что и выше, через `Optional`...

CbaKKa commented on bot/src/main/java/edu/java/bot/commands/CommandTrack.java:
```diff
@@ -0,0 +1,54 @@
+    private String prepareTrackMessage(Long chatId) {
+        var userOptional = userServise.findUserById(chatId);
+
+        if (userOptional.isPresent()) {
```
И тут

CbaKKa commented on bot/src/main/java/edu/java/bot/controller/MessageController.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @Override
+    public int process(List<Update> list) {
```
Желательно вынести в отдельные приватные методы для большей читабельности

Mur0riki committed:
edit Command classes

CbaKKa commented on bot/src/main/java/edu/java/bot/commands/CommandList.java:
```diff
@@ -0,0 +1,52 @@
+    @Override public String handle(Update update) {
+        var chatId = update.message().chat().id();
+
+        var userOptional = userService.findUserById(chatId);
```
Не совсем правильно используется `Optional`

CbaKKa commented on bot/src/main/java/edu/java/bot/commands/CommandList.java:
```diff
@@ -0,0 +1,52 @@
+                : prepareListSitesMessage(userOptional.get().getSites())).orElse(UNKNOWN_USER);
+    }
+
+    private String prepareListSitesMessage(List<URI> uriList) {
```
Можно сделать красивее: переменная не нужна, в конструктор `StringBuilder` можно передать USER_TRACK_SITES_MESSAGE, а не через `.append()`

CbaKKa commented on bot/src/main/java/edu/java/bot/commands/CommandStart.java:
```diff
@@ -0,0 +1,48 @@
+
+    private String registerUser(long chatId) {
+        Optional userOptional = userService.findUserById(chatId);
+        return (String) userOptional.map(u -> ALREADY_REGISTRATE_MESSAGE)
```
Так же неправильное использование `Optional`:

```java
        return userService.findUserById(chatId)
                          .map(u -> ALREADY_REGISTRATE_MESSAGE)
                          .orElseGet(() -> {
                              userService.saveUser(new User(chatId, List.of(), SessionState.BASE_STATE));

                              return SUCCESS_REGISTRATION_MESSAGE;
                          });
```

CbaKKa reviewed (changes requested):
Привет, у меня сложилось впечатление, что ты не до конца понимаешь как работает `Optional` и для чего он нужен, можем созвониться, либо самостоятельно изучи этот вопрос, ответь, пожалуйста, до вечера 20.02

Mur0riki commented:
Я бы хотел как изучить самостоятельно, так и созвониться после самостоятельного изучения для понимания того насколько я правильно понял Optional и для чего он нужен.

CbaKKa commented:
Тогда напиши мне в тг @max_zhurov, созвонимся как будешь готов

Mur0riki committed:
edit Command classes
Corrected variable names
edut User class
delete @Qualifier Annotation
edit Command interface(replace Update update in method handle to long chatId)
make User class more encapsulated

CbaKKa commented on bot/src/main/java/edu/java/bot/commands/CommandTrack.java:
```diff
@@ -0,0 +1,46 @@
+        return "Позволяет начать отслеживать в нашем боте нужный вам сайт.";
+    }
+
+    @Override
```
Нужно навесить `@Transactional` т.к. нужно обеспечить консистентность данных (пока читаем и ставим `state` что-то может измениться)

CbaKKa commented on bot/src/main/java/edu/java/bot/commands/CommandTrack.java:
```diff
@@ -0,0 +1,46 @@
+    private String prepareTrackMessage(Long chatId) {
+        return userService.findUserById(chatId).map(
+            user -> {
+                changeStatusUserAndSave(userService.findUserById(chatId).get());
```
Надо переделать на что-то такое: `userService.findUserById(chatId).ifPresent(this::changeStatusUserAndSave)`

Mur0riki replied:
Мне кажется тут более удобно будет использовать Supplier нежели Consumer потому что мне нужно возвращать String исходя из того найден пользователь в бд или нет

CbaKKa replied:
Если код дошел до этого места - то он 100% найден.
Вообще, повторный вызов репозитория здесь не нужен

CbaKKa replied:
Имеет смысл добавить `.map(User::getId).map(...)` и внутри второго мапа уже вызывать этот приватный метод

CbaKKa commented on bot/src/main/java/edu/java/bot/commands/CommandUntrack.java:
```diff
@@ -0,0 +1,46 @@
+    private String prepareUnTrackMessage(long chatId) {
+        return userService.findUserById(chatId).map(
+            user -> {
+                changeStatusUserAndSave(userService.findUserById(chatId).get());
```
Тут тоже надо `ifPresent`

CbaKKa commented on bot/src/main/java/edu/java/bot/service/MessageService.java:
```diff
@@ -0,0 +1,113 @@
+@Service
+public class MessageService {
+    public static final String DO_REGISTRATION_MESSAGE = "Необходимо зарегистрироваться.";
+    public static final String INVALID_URI_MESSAGE = "Неверно указан URI.";
```
Лучше удалять неиспользуемые переменные, потом, если что, добавить заново

CbaKKa commented on bot/src/main/java/edu/java/bot/service/MessageService.java:
```diff
@@ -0,0 +1,113 @@
+    private String processNonCommandMessage(Long chatId, String text) {
+        return userRepository.findUserById(chatId).map(u -> {
+            try {
+                return processStateUserMessage(userRepository.findUserById(chatId).get(), new URI(text));
```
Переделать на `ifPresent`

CbaKKa reviewed (changes requested):
В целом, все отлично, пару мелочей поправить надо

Mur0riki committed:
delete unused variable
add Annotation @Transactional to CommandTrack and CommandUntrack

Mur0riki committed:
remove unused import and @BeforeClass Annotation from CommandTest
Add Annotation @Transactional on registerUser method in CommandStart
edit CommandTrack
edit CommandUntrack
